### PR TITLE
SDK-11 [Java SDK] Update Readme.md file

### PR DESCRIPTION
**Pull request by VladimirLevitskiy**
Pull Request Checklist
======================

Please read this for our PR Culture:
------------------------------------
https://incountry.atlassian.net/wiki/spaces/ED/pages/46858869/Pull+requests+culture+Draft

Work relates to:
----------------
Link(s) to Jira Task/User Story/Bug/Epic:
- https://incountry.atlassian.net/browse/SDK-11

Code Quality (Before submitting PR)
------------
- [x] Ensure code compiles and passes Sonar Lint.
- [x] Do your best to ensure integration tests pass. If they don't then...
- [x] Remove, ignore, or mark pending any brittle tests.
- [x] Remove commented out code.

Target Release
--------------
- [ ] Please put the target release vehicle this needs to be bundled with.
- [ ] Did you increment the version? What is the new version?

### CODE REVIEW

File: pull_request_template.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ https://incountry.atlassian.net/wiki/spaces/ED/pages/46858869/Pull+requests+cult
 Work relates to:
 ----------------
 Link(s) to Jira Task/User Story/Bug/Epic:
-- [ ]
+- [ https://incountry.atlassian.net/browse/SDK-XX ]
```
Comment (RommeroInCountry) [inline pull_request_template.md:11]: Better use this one 
- [SDK-XX](https://incountry.atlassian.net/browse/SDK-XX)

File: src/integration/java/com/incountry/residence/sdk/FullMigrationExample.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+                "    \"secrets\": [\n" +
+                "        {\"secret\": \"password0\", \"version\": 0},\n" +
+                "        {\"secret\": \"password1\", \"version\": 1},\n" +
+                "    ],\n" +
+                "}";
+        SecretKeyAccessor secretKeyAccessor = SecretKeyAccessor.getAccessor(() -> secretsDataInJson);
+        String endPoint = "https://us.api.incountry.example";
+        String country = "US";
+        String envId = "someEnvironmentId";
+        String apiKey = "someApyKey";
```
Comment (RommeroInCountry) [inline src/integration/java/com/incountry/residence/sdk/FullMigrationExample.java:21]: Api, we are not in Alladin :D

File: src/main/java/com/incountry/residence/sdk/StorageImpl.java [hunk 3]
```diff
@@ -50,20 +50,42 @@ private static String loadFromEnv(String key) {
     }
 
     private StorageImpl() {
-
     }
 
+    /**
+     * creating Storage instance with ENV variables without encryption (for debugging)
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/StorageImpl.java:56]: It's not actually for debugging but more for client-side encryption

File: src/main/java/com/incountry/residence/sdk/StorageImpl.java [hunk 4]
```diff
@@ -50,20 +50,42 @@ private static String loadFromEnv(String key) {
...
     public static Storage getInstance() throws StorageServerException {
         return getInstance(null);
     }
 
+    /**
+     * creating Storage instance with ENV variables
+     *
+     * @param secretKeyAccessor Instance of SecretKeyAccessor class. Used to fetch encryption secret
+     * @return instance of Storage
+     * @throws StorageServerException when configuration is invalid
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/StorageImpl.java:70]: StorageServerException is for http connections/PoPAPI errors. If smth is wrong with configuration (based on params user provided) it's more likely to be StorageClientException.
Either the description is not correct or the exception should be changed I think

File: src/main/java/com/incountry/residence/sdk/dto/Record.java [hunk 5]
```diff
@@ -28,12 +28,12 @@ public Record(String key, String body) {
...
-     * @param key        key
-     * @param body       data to be stored and encrypted
-     * @param profileKey profile key
-     * @param rangeKey   range key for sorting in pagination
-     * @param key2       key2
-     * @param key3       key3
+     * @param key        Required, record key
+     * @param body       Optional, data to be stored and encrypted
+     * @param profileKey Optional, profile key
+     * @param rangeKey   Optional, range key for sorting in pagination
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/dto/Record.java:34]: Range key have more applications, not only sorting and pagination :) it's just a numerical key which allows comparison

File: README.md [hunk 6]
```diff
@@ -3,289 +3,532 @@ InCountry Storage SDK
...
+SecretsData JSON object
+```json
+{
+  "secrets": [{
+       "secret": "someSecret",  // String   
+       "version": "0",          // Should be a non-negative integer
+       "isKey": "true"          // Boolean, should be 'true' only for user-defined encryption keys
     }
+  ],
+  "currentVersion": "0"         // Should be a non-negative integer. One of the secrets must have same version as currentVersion in SecretsData
```
Comment (RommeroInCountry) [inline README.md:127]: version, currentVersion should be plain integers (not strings)
isKey should be just true, not "true"

File: README.md [hunk 7]
```diff
@@ -3,289 +3,532 @@ InCountry Storage SDK
...
+                .rangeKeyEq(someRnageKey)
+                .limitAndOffset(50, 0);
 ```
 
 Here is the example of how `find` method can be used:
-
+```java
+FindFilterBuilder builder = FindFilterBuilder.create()
+                .key2Eq("kitty")
+                .key3NotIn(Arrays.asList("bow-wow","сock-a-doodle-do"))
```
Comment (RommeroInCountry) [inline README.md:389]: Let's use smth more simple like "<key3_1>", "<key3_2>" :D :D :D

File: README.md [hunk 8]
```diff
@@ -3,289 +3,532 @@ InCountry Storage SDK
...
+    try {
+        // use InCountry Storage instance here
+    } catch (StorageClientException e) {
+        // some input validation error
+    } catch (StorageServerException e) {
+        // some server error
+    } catch (StorageCryptoException e) {
+        // some encryption error
+    } catch (StorageException e) {
+        // some input validation error
```
Comment (RommeroInCountry) [inline README.md:529]: general error, not input validation

File: src/main/java/com/incountry/residence/sdk/StorageImpl.java [hunk 9]
```diff
@@ -53,7 +53,7 @@ private StorageImpl() {
     }
 
     /**
-     * creating Storage instance with ENV variables without encryption (for debugging)
+     * creating Storage instance with ENV variables without encryption at SDK
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/StorageImpl.java:56]: let's simply go with "Storage instance with ENV variables without encryption"

File: src/main/java/com/incountry/residence/sdk/StorageImpl.java (hunk 4 above)
Comment (VladimirLevitskiy) [reply src/main/java/com/incountry/residence/sdk/StorageImpl.java:70]: Fix partly. Full fix will be in task SDK-79

File: README.md [hunk 10]
```diff
@@ -3,289 +3,552 @@ InCountry Storage SDK
...
+}
+```
 
-        SecretsData secretsData = new SecretsData();
-        secretsData.setSecrets(secretKeyList);
-        secretsData.setCurrentVersion(0);
-        return secretsData;
+3. As an object implementing `SecretsDataGenerator` interface. SecretsDataGenerator's `generate` method should return `SecretsData` object or String. String can be simple key or a valid JSON string. This JSON string will then be parsed as a `SecretsData`
+SecretsData JSON object
+```json
```
Comment (RommeroInCountry) [inline README.md:131]: let's change it to javascript so that the comments don't look red and ugly

**Pull request merged**